File: here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/models/features/ExtensionConfig.java
/*
 * Copyright (C) 2017-2023 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.naksha.lib.core.models.features;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.here.naksha.lib.core.models.geojson.implementation.XyzFeature;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@JsonTypeName
public class ExtensionConfig extends XyzFeature {

  public static final String EXPIRY = "expiry";
  public static final String EXTENSIONS = "extensions";
  public static final String WHITELIST_DELEGATE_CLASS = "whitelistDelegateClass";

  @JsonProperty(EXPIRY)
  long expiry;

  @JsonProperty(EXTENSIONS)
  List<Extension> extensions;

  @JsonProperty(WHITELIST_DELEGATE_CLASS)
  List<String> whitelistDelegateClass;

  public ExtensionConfig(
      @JsonProperty(EXPIRY) @NotNull long expiry,
      @JsonProperty(EXTENSIONS) @NotNull List<Extension> extensions,
      @JsonProperty(WHITELIST_DELEGATE_CLASS) @Nullable List<String> whitelistDelegateClass) {
    this.expiry = expiry;
    this.extensions = extensions;
    this.whitelistDelegateClass = whitelistDelegateClass;
  }

  public long getExpiry() {
    return expiry;
  }

  public void setExpiry(long expiry) {
    this.expiry = expiry;
  }

  public List<Extension> getExtensions() {
    return extensions;
  }

  public void setExtensions(List<Extension> extensions) {
    this.extensions = extensions;
  }

  public List<String> getWhilelistDelegateClass() {
    return whitelistDelegateClass;
  }

  public void setWhilelistDelegateClass(List<String> whitelistDelegateClass) {
    this.whitelistDelegateClass = whitelistDelegateClass;
  }
}

File: here-naksha-lib-core/src/jvmMain/kotlin/com/here/naksha/lib/core/models/ExtensionConfig.kt
package com.here.naksha.lib.core.models

import com.here.naksha.lib.core.models.features.Extension
import naksha.base.AnyObject
import naksha.base.NotNullProperty
import naksha.base.NullableProperty
import naksha.base.StringList

class ExtensionConfig() : AnyObject() {
    companion object {
        private val EXTENSIONS_NULL = NullableProperty<ExtensionConfig, ExtensionList>(ExtensionList::class)
        private val WHITELIST_DELEGATE_CLASSES_NULL = NullableProperty<ExtensionConfig, StringList>(StringList::class)
        private val LONG = NotNullProperty<ExtensionConfig, Long>(Long::class)
    }

    var expiry by LONG
    var extensions by EXTENSIONS_NULL
    var whitelistDelegateClasses by WHITELIST_DELEGATE_CLASSES_NULL

    constructor(
        expiry: Long,
        extensions: List<Extension>?=null,
        whitelistDelegateClasses: List<String>?=null
    ) : this() {
        this.expiry = expiry
        this.extensions = extensions?.let(ExtensionList::fromList)
        this.whitelistDelegateClasses = whitelistDelegateClasses?.let(StringList::fromList)
    }
}